### PR TITLE
rkt: Support multiple containers in nomad

### PR DESCRIPTION
**Pull request by indradhanush**
The job spec for the rkt driver now supports an additional key:
`pod_manifest`; This makes `image` and `pod_manifest` exclusive of each
other. Currently the pod_manifest supplied as user input can be
accepted if it does not have an image id. rkt fetches the image using
the image name and modifies the pod manifest with the obtained image
id before passing it along to `rkt prepare`. 

As a result, now multiple rkt containers can run within the same nomad 
task.

**Testing:**
- Setup nomad locally.
- `sudo consul agent -dev` 
- `make dev && sudo bin/nomad agent -dev`.
- Get [example.nomad](https://gist.github.com/indradhanush/f0747d8acc42e353eeeb36c05d76e457).
- `bin/nomad run example.nomad`.
- To test a pod with multiple containers get: [dual.nomad](https://gist.github.com/indradhanush/57d86273d499522ad29e87f497b7922e) and run `bin/nomad run dual.nomad`.

**Expected output:**
- `bin/nomad status example` should give you a status of `running` after a few minutes.
- `bin/nomad alloc-status <alloc-id>` should give the status of redis.
- `bin/nomad status dual` should return the `alloc-id`. Use it to get the `alloc-status`. This should tell you the ports on which `nginx` and `redis` are running. The port named `server` runs nginx and you can test it with `curl http://127.0.0.1:<port>`. This should return the default nginx page.
- Get the [redis cli](https://redis.io/clients) and run it with `redis-cli -p <port>` and play around. :)

### CODE REVIEW

File: client/driver/rkt.go [hunk 1]
```diff
@@ -117,6 +120,28 @@ type rktPID struct {
 	MaxKillTimeout time.Duration
 }
 
+type appcRuntimeApp struct {
+	lastditch.RuntimeApp
```
Comment (krnowak) [inline client/driver/rkt.go]: What was the reason for using lastditch instead of appctypes here? It's confusing a bit, since you use appctypes for the rest of the fields here. Some comment explaining it would be great.

File: client/driver/rkt.go [hunk 2]
```diff
@@ -435,66 +472,181 @@ func (d *RktDriver) Start(ctx *ExecContext, task *structs.Task) (*StartResponse,
...
+		// Add memory isolator
+		prepareArgs = append(prepareArgs, fmt.Sprintf("--memory=%vM", int64(task.Resources.MemoryMB)))
 
-	// Add memory isolator
-	prepareArgs = append(prepareArgs, fmt.Sprintf("--memory=%vM", int64(task.Resources.MemoryMB)))
+		// Add CPU isolator
+		prepareArgs = append(prepareArgs, fmt.Sprintf("--cpu=%vm", int64(task.Resources.CPU)))
+
+	} else {
+		allocACName, err := appctypes.NewACName(allocVolName)
```
Comment (krnowak) [inline client/driver/rkt.go]: Where does `allocVolName` come from? If this is entirely controlled by program then maybe you could just use `appctypes.MustACName(allocVolName)`? Same for `localVolName` and `secretsVolName`.

Comment (krnowak) [reply client/driver/rkt.go]: If this can't be done, then please use better error messages. Something like `fmt.Errorf("[ERR] driver.rkt: failed to convert %s to ACName for pod manifest for /alloc volume: %v", allocVolName, err)` or something.

File: client/driver/rkt.go [hunk 3]
```diff
@@ -435,66 +472,181 @@ func (d *RktDriver) Start(ctx *ExecContext, task *structs.Task) (*StartResponse,
...
+
+		var environment appctypes.Environment
+		for name, value := range ctx.TaskEnv.Map() {
+			environment = append(environment, appctypes.EnvironmentVariable{
+				Name:  name,
+				Value: value,
+			})
+		}
+
+		fetchArgs := make([]string, 0, 50)
```
Comment (krnowak) [inline client/driver/rkt.go]: The `fetchArgs` variable could be created inside the `fetchImage` function. So `fetchImage` could just take globalArgs. The reason is that `fetchImage` knows exactly how large the array can be, so you could avoid using some magic number like 50, and `fetchImage` can add the `"fetch"` arg itself.

File: client/driver/rkt.go [hunk 4]
```diff
@@ -435,66 +472,181 @@ func (d *RktDriver) Start(ctx *ExecContext, task *structs.Task) (*StartResponse,
...
+			}
+
+			if isolators.GetByName(*memoryACIdentifier) == nil {
+				s := fmt.Sprintf("%vM", task.Resources.MemoryMB)
+				resMemory, err := appctypes.NewResourceMemoryIsolator(s, s)
+				if err != nil {
+					return nil, fmt.Errorf("[ERR] driver.rkt: failed to create ResourceMemory: %s", err)
+				}
+
+				currApp.App.Isolators = append(isolators, resMemory.AsIsolator())
```
Comment (krnowak) [inline client/driver/rkt.go]: You could just do `isolators = append(isolators, resMemory.AsIsolator())` here and for the CPU isolator, and after adding all the isolators do `currApp.App.Isolators = isolators`.

File: client/driver/rkt.go [hunk 5]
```diff
@@ -435,66 +472,181 @@ func (d *RktDriver) Start(ctx *ExecContext, task *structs.Task) (*StartResponse,
...
+			}
+
+			// Add CPU isolator
+			isolators = currApp.App.Isolators
```
Comment (krnowak) [inline client/driver/rkt.go]: With the above change, this line could be dropped.

File: client/driver/rkt.go (hunk 3 above)
Comment (krnowak) [reply client/driver/rkt.go]: Another reason is that `fetchImage` modifies `fetchArgs` so it gets changed after each iteration. And I don't think that `append` function makes a deepcopy of the modified slice. That looks fishy.

File: client/driver/rkt.go [hunk 6]
```diff
@@ -731,6 +951,36 @@ func (d *RktDriver) Open(ctx *ExecContext, handleID string) (DriverHandle, error
 	return h, nil
 }
 
+func (d *RktDriver) fetchImage(img lastditch.RuntimeImage, fetchArgs []string) (string, error) {
```
Comment (krnowak) [inline client/driver/rkt.go]: So this function could take globalArgs instead and do `fetchArgs := make([]String, 0, len(globalArgs) + 2 /* one for "fetch", one for image name */)`.

File: client/driver/rkt.go [hunk 7]
```diff
@@ -435,66 +472,181 @@ func (d *RktDriver) Start(ctx *ExecContext, task *structs.Task) (*StartResponse,
...
+
+			d.logger.Printf("[DEBUG] driver.rkt: Image ID obtained: %s", ID)
+
+			currApp.Image.ID = ID
+			currApp.Mounts = mounts
+			currApp.App.Environment = append(currApp.App.Environment, environment...)
+
+			// Add memory isolator
+			isolators := currApp.App.Isolators
+			memoryACIdentifier, err := appctypes.NewACIdentifier(appctypes.ResourceMemoryName)
```
Comment (krnowak) [inline client/driver/rkt.go]: You probably could use `MustACIdentifier` here.

File: client/driver/rkt.go [hunk 8]
```diff
@@ -435,66 +472,181 @@ func (d *RktDriver) Start(ctx *ExecContext, task *structs.Task) (*StartResponse,
...
+				if err != nil {
+					return nil, fmt.Errorf("[ERR] driver.rkt: failed to create ResourceMemory: %s", err)
+				}
+
+				currApp.App.Isolators = append(isolators, resMemory.AsIsolator())
+			}
+
+			// Add CPU isolator
+			isolators = currApp.App.Isolators
+			cpuACIdentifier, err := appctypes.NewACIdentifier(appctypes.ResourceCPUName)
```
Comment (krnowak) [inline client/driver/rkt.go]: You probably could use `MustACIdentifier` here.

File: client/driver/rkt.go [hunk 9]
```diff
@@ -545,11 +697,23 @@ func (d *RktDriver) Start(ctx *ExecContext, task *structs.Task) (*StartResponse,
...
+
 			// Add port option to rkt run arguments. rkt allows multiple port args
-			prepareArgs = append(prepareArgs, fmt.Sprintf("--port=%s:%s", containerPort, hostPortStr))
+			if img != "" {
+				hostPortStr := strconv.Itoa(port.Value)
+				prepareArgs = append(prepareArgs, fmt.Sprintf("--port=%s:%s", containerPort, hostPortStr))
+			} else {
+				containerPortACName, err := appctypes.NewACName(containerPort)
+				if err != nil {
+					return nil, fmt.Errorf("[ERR] driver.rkt: failed to set ACName: %s", err)
```
Comment (krnowak) [inline client/driver/rkt.go]: `"… failed to set ACName for %s: %v", containerPort, err)`?

File: client/driver/rkt.go [hunk 10]
```diff
@@ -563,13 +727,56 @@ func (d *RktDriver) Start(ctx *ExecContext, task *structs.Task) (*StartResponse,
...
```
Comment (krnowak) [inline client/driver/rkt.go]: Same here.

**Review by krnowak — CHANGES_REQUESTED** [review]
That's a bit of a "unsolicited" review, but anyway, I needed a break. ;)

The `start` function is a 500 lines monstrosity, ew. But I don't think changing this is in scope of this work.

Had some nits here, likely nothing serious, but I only checked the code.

Not sure what is the convention here for error reporting, but I remember using usually `%v` parameter for formatting errors, not `%s`.

File: client/driver/rkt.go (hunk 1 above)
Comment (iaguis) [reply client/driver/rkt.go]: Regular appc types require ID to be set, but the manifest we take doesn't have it. I agree a comment is nice.

File: client/driver/rkt.go [hunk 11]
```diff
@@ -731,6 +951,36 @@ func (d *RktDriver) Open(ctx *ExecContext, handleID string) (DriverHandle, error
 	return h, nil
 }
 
+func (d *RktDriver) fetchImage(img lastditch.RuntimeImage, fetchArgs []string) (string, error) {
+	var version string
```
Comment (iaguis) [inline client/driver/rkt.go]: Can we set `version` to `latest` by default and use that if there's no version specified instead of erroring out?

File: client/driver/rkt.go [hunk 12]
```diff
@@ -731,6 +951,36 @@ func (d *RktDriver) Open(ctx *ExecContext, handleID string) (DriverHandle, error
...
+			break
+		}
+	}
+
+	var imgName string
+	if version == "" {
+		return "", fmt.Errorf("no version supplied for image: %s", img.Name)
+	}
+
+	imgName = fmt.Sprintf("docker://%s:%s", img.Name, version)
```
Comment (iaguis) [inline client/driver/rkt.go]: If we do this, we'd only support docker images (converted via `docker2aci`) so regular ACI images won't be supported. For example: `rkt run coreos.com/etcd:v2.0.10` works, but if we use `coreos.com/etcd` in a pod manifest, it will try to fetch `docker://coreos.com/etcd`, which will fail.

The problem is that we can't use the syntax `docker://` in `imgName` with the regular appc schema PodManifest. However, we're now using `lastditch` so that should be possible right?

If that's possible, I suggest deviating again from the appc spec and accept image names like `docker://`, passing that directly to rkt.

File: client/driver/rkt.go [hunk 13]
```diff
@@ -563,13 +727,56 @@ func (d *RktDriver) Start(ctx *ExecContext, task *structs.Task) (*StartResponse,
...
+		}
+
+		// Ensure that the generated pod manifest is valid according to the appc spec
+		validatePodManifest := appcschema.PodManifest{}
+		if err := json.Unmarshal(podManifestStr, &validatePodManifest); err != nil {
+			return nil, fmt.Errorf("[ERR] driver.rkt: Invalid rkt pod manifest: %s", err)
+		}
+
+		// Create temporary file to dump the contents of podManifest
+		tmpfile, err := ioutil.TempFile("", "pod")
```
Comment (iaguis) [inline client/driver/rkt.go]: We should namespace this a bit better, how about:

```
ioutil.TempFile("", "nomad-rkt-pod")
```

File: client/driver/rkt.go [hunk 14]
```diff
@@ -563,13 +727,56 @@ func (d *RktDriver) Start(ctx *ExecContext, task *structs.Task) (*StartResponse,
...
+		if err != nil {
+			return nil, err
+		}
+
+		podManifestTmpfile := tmpfile.Name()
+		defer os.Remove(podManifestTmpfile)
+		defer tmpfile.Close()
```
Comment (iaguis) [inline client/driver/rkt.go]: I'd avoid deferring the `Close()` call and do it as a regular function call after writing to the file so we make sure the file is closed before rkt uses it and we catch possible errors when closing.

**Review by iaguis — CHANGES_REQUESTED** [review]
Some comments.

File: client/driver/rkt.go [hunk 15]
```diff
@@ -257,9 +282,13 @@ func (d *RktDriver) Validate(config map[string]interface{}) error {
 	fd := &fields.FieldData{
 		Raw: config,
 		Schema: map[string]*fields.FieldSchema{
+			// image is only required if pod_manifest was not passed in
```
Comment (lilic) [inline client/driver/rkt.go:291]: Please make all your comments full sentences as per golang preferred guide, unless the project style guides does not recommend that?

File: helper/fields/data.go [hunk 16]
```diff
@@ -18,7 +18,16 @@ type FieldData struct {
...
 	var result *multierror.Error
 
-	// Scan for missing required fields
+	// Check that either image or pod_manifest was passed in but not both or not at all
+	_, podManifestExists := d.Raw["pod_manifest"]
+	_, imageExists := d.Raw["image"]
+
+	if podManifestExists == imageExists {
+		result = multierror.Append(result, fmt.Errorf(
+			"Either field image or field pod_manifest is required but not both"))
```
Comment (lilic) [inline helper/fields/data.go]: I am a bit confused by this? Not sure this works, as if they are both false this if statment would pass as well?

Maybe `if podManifestExists  && imageExists  {...}` would be a bit more clear.

Also I would rephrase the error msg to something like:
> Both field image and pod manifest were passed. Only one is required.
And also print out the values of both vars.

**Review by lilic — CHANGES_REQUESTED** [review]
Just some things that caught my eye.

File: helper/fields/data.go (hunk 16 above)
Comment (krnowak) [reply helper/fields/data.go]: You need to pass either an image or a pod manifest. Passing nothing or passing both is an error.

Comment (lilic) [reply helper/fields/data.go]: Then just the error msg should be reworked.

File: client/driver/rkt.go (hunk 3 above)
Comment (indradhanush) [reply client/driver/rkt.go]: My understanding is that `fetchImage` gets a new reference to the newly created slice while the line:

```
ID, err := d.fetchImage(app.Image, fetchArgs)
```

always refers to the `fetchArgs` created outside the for loop. As a result the state of the `fetchArgs` remains unaltered after each iteration.

Here's a snippet for validation: https://play.golang.org/p/0kq6tCLGk6d

However I agree that we can declare `fetchArgs` inside the function itself and get rid of the magic number.

Comment (krnowak) [reply client/driver/rkt.go]: The slices are a bit tricky and I don't remember all the details. So in this case it seems to work fine as long as we don't modify the contents that were already in the slice. Consider this: https://play.golang.org/p/xvLJbVqdd48

File: helper/fields/data.go (hunk 16 above)
Comment (indradhanush) [reply helper/fields/data.go]: @LiliC Fixed the error message as suggested but not printing out the variables. It would be a lot of noise with the `pod_manifest` being potentially huge, don't you think?

File: helper/fields/data.go [hunk 17]
```diff
@@ -18,7 +18,16 @@ type FieldData struct {
...
 	var result *multierror.Error
 
-	// Scan for missing required fields
+	// Check that either image or pod_manifest was passed in but not both or not at all
+	_, podManifestExists := d.Raw["pod_manifest"]
+	_, imageExists := d.Raw["image"]
+
+	if podManifestExists == imageExists {
+		result = multierror.Append(result, fmt.Errorf(
+			"Both field image and pod_manifest were passed. Only one is required."))
```
Comment (lilic) [inline helper/fields/data.go]: @indradhanush Think including the following part is helpful:

> "Both field image and pod_manifest were passed. Either specify one or no fields."

File: client/driver/rkt.go [hunk 18]
```diff
@@ -208,19 +239,18 @@ func rktManifestMakePortMap(manifest *appcschema.PodManifest, configPortMap map[
...
-	}
 
 	portMap := make(map[string]int)
-	for svc, name := range configPortMap {
-		for _, port := range app.App.Ports {
-			if port.Name.String() == name {
-				portMap[svc] = int(port.Port)
+	for _, app := range manifest.Apps {
+		if app.App == nil {
+			return nil, fmt.Errorf("specified app %s has no App object", app.Name)
```
Comment (iaguis) [inline client/driver/rkt.go]: nit: use `%q` instead of `%s`

File: client/driver/rkt.go [hunk 19]
```diff
@@ -384,6 +419,14 @@ func (d *RktDriver) Start(ctx *ExecContext, task *structs.Task) (*StartResponse,
 	// ACI image
 	img := driverConfig.ImageName
 
+	// Check if the pod manifest passed in is valid
```
Comment (lilic) [inline client/driver/rkt.go]: Everywhere it seems like you are missing a `.` at the end of the sentence. I checked and it seems like they are going by that style, so I would stick to that.

File: client/driver/rkt.go [hunk 20]
```diff
@@ -435,66 +478,139 @@ func (d *RktDriver) Start(ctx *ExecContext, task *structs.Task) (*StartResponse,
 	// Convert underscores to dashes in task names for use in volume names #2358
 	sanitizedName := strings.Replace(task.Name, "_", "-", -1)
 
-	// Mount /alloc
+	// Format volume names appropriately
```
Comment (lilic) [inline client/driver/rkt.go]: Confused by this comment. :)

File: client/driver/rkt.go [hunk 21]
```diff
@@ -563,18 +691,64 @@ func (d *RktDriver) Start(ctx *ExecContext, task *structs.Task) (*StartResponse,
...
+		// If a user has been specified for the task, pass it through to the user
+		if task.User != "" {
+			prepareArgs = append(prepareArgs, fmt.Sprintf("--user=%s", task.User))
+		}
 	}
 
-	// If a user has been specified for the task, pass it through to the user
-	if task.User != "" {
-		prepareArgs = append(prepareArgs, fmt.Sprintf("--user=%s", task.User))
+	if img == "" {
```
Comment (lilic) [inline client/driver/rkt.go]: I think this if statement will always be true. Or am I missing something? :)

File: client/driver/rkt.go [hunk 22]
```diff
@@ -731,6 +926,77 @@ func (d *RktDriver) Open(ctx *ExecContext, handleID string) (DriverHandle, error
 	return h, nil
 }
 
+func (d *RktDriver) fetchImage(img lastditch.RuntimeImage, globalArgs []string) (string, error) {
+	var version string
```
Comment (lilic) [inline client/driver/rkt.go]: You can just do version := "latest" and then override if set like you do below. WDYT?

File: client/driver/rkt.go [hunk 23]
```diff
@@ -731,6 +926,77 @@ func (d *RktDriver) Open(ctx *ExecContext, handleID string) (DriverHandle, error
...
+			break
+		}
+	}
+
+	// If no version is supplied, set it to latest
+	if version == "" {
+		version = "latest"
+	}
+
+	imgName := fmt.Sprintf("%s:%s", img.Name, version)
```
Comment (lilic) [inline client/driver/rkt.go]: What if `img.Name` is empty?

File: vendor/github.com/appc/spec/schema/lastditch/pod.go [hunk 24]
```diff
@@ -0,0 +1,57 @@
...
+type RuntimeImage struct {
+	Name   string `json:"name,omitempty"`
+	ID     string `json:"id"`
+	Labels Labels `json:"labels,omitempty"`
+}
+
+// a type just to avoid a recursion during unmarshalling
+type podManifest PodManifest
+
+func (pm *PodManifest) UnmarshalJSON(data []byte) error {
```
Comment (lilic) [inline vendor/github.com/appc/spec/schema/lastditch/pod.go:46]: All functions that a re exported should have a comment like below:
```
// UnmarshalJSON unmarshals JSON data....
func (pm *PodManifest) UnmarshalJSON(data []byte) error {
....
``

File: client/driver/rkt_test.go [hunk 25]
```diff
@@ -139,7 +139,155 @@ func TestRktDriver_Start_DNS(t *testing.T) {
...
+	if _, err := d.Prestart(ctx.ExecCtx, task); err != nil {
+		t.Fatalf("error in prestart: %v", err)
+	}
+	resp, err := d.Start(ctx.ExecCtx, task)
+	if err != nil {
+		t.Fatalf("err: %v", err)
+	}
+	defer resp.Handle.Kill()
+
+	// Attempt to open
```
Comment (lilic) [inline client/driver/rkt_test.go]: open what? :)

**Review by lilic — CHANGES_REQUESTED** [review]
Just had a quick look. Left some suggestions.

File: client/driver/rkt.go (hunk 19 above)
Comment (indradhanush) [reply client/driver/rkt.go]: A bit confused about which style you meant, i.e.: to use dots at the end or not? In places that I checked, the project was not using dots to end sentences.

File: client/driver/rkt.go (hunk 21 above)
Comment (indradhanush) [reply client/driver/rkt.go]: This can be false if the pod manifest was supplied by the user instead.

File: vendor/github.com/appc/spec/schema/lastditch/pod.go (hunk 24 above)
Comment (indradhanush) [reply vendor/github.com/appc/spec/schema/lastditch/pod.go:46]: This was vendored from [here](https://github.com/appc/spec/blob/master/schema/lastditch/pod.go#L46). Does it qualify for this PR?

File: client/driver/rkt_test.go (hunk 25 above)
Comment (indradhanush) [reply client/driver/rkt_test.go]: This comment appears to have been copied over from other tests that do something similar: https://github.com/hashicorp/nomad/blob/master/client/driver/rkt_test.go#L131 ; While I agree this comment doesn't explain what it's attempting to open, but should we try to address that here or stick with what upstream already has?

File: client/driver/rkt.go (hunk 21 above)
Comment (lilic) [reply client/driver/rkt.go]: Well right above you check `if img != "" {...}`. I find this handling of conditions a bit awkward and not so clear. IMHO.

File: vendor/github.com/appc/spec/schema/lastditch/pod.go (hunk 24 above)
Comment (lilic) [reply vendor/github.com/appc/spec/schema/lastditch/pod.go:46]: Ooops, sorry missed the `vendor` part. Ignore.

File: client/driver/rkt_test.go (hunk 25 above)
Comment (lilic) [reply client/driver/rkt_test.go]: Up to you I guess. If you found it confusing, it doesn't hurt to help the next person understand it better.

File: client/driver/rkt.go (hunk 19 above)
Comment (lilic) [reply client/driver/rkt.go]: They seem to have missed style. For documentation it always needs to be a full sentence, but people tend to stick with all comments being full sentences, up to you.

File: client/driver/rkt.go [hunk 26]
```diff
@@ -764,12 +1030,22 @@ func (h *rktHandle) Exec(ctx context.Context, cmd string, args []string) ([]byte
 	if h.uuid == "" {
 		return nil, 0, fmt.Errorf("unable to find rkt pod UUID")
 	}
+
 	// enter + UUID + cmd + args...
-	enterArgs := make([]string, 3+len(args))
+	// If a pod manifest with multiple apps is used, use the first app for the command:
+	// enter + --<app=appname> + UUID + cmd + args...
```
Comment (kosyfrances) [inline client/driver/rkt.go:1036]: @indradhanush IMO, Having a `TO DO` explanation here like I did in my branch, of some better way to fix this hack will be very helpful. We know this is a hack, and the ideal thing is to let the required app's name be passed in, rather than just using the first app. I think skipping that explanation will not help us or them remember that the source of that code needs to be adjusted for it to be flexible when a multiple apps PodManifest is passed in.

Comment (indradhanush) [reply client/driver/rkt.go:1036]: @kosyfrances Yes this sounds good to me. I'm not very sure why I had removed it back while rebasing. Sorry about that.

File: client/driver/rkt.go [hunk 27]
```diff
@@ -117,6 +121,34 @@ type rktPID struct {
 	MaxKillTimeout time.Duration
 }
 
+// The appc spec makes the image ID mandatory. However while passing
+// the pod_manifest in the nomad job spec, the ID will not be
+// available. As a result we use lastditch.RuntimeApp to skip
+// validation on the ID field initially and fallback to the appc
```
Comment (iaguis) [inline client/driver/rkt.go]: fallback -> fall back

File: client/driver/rkt.go [hunk 28]
```diff
@@ -257,9 +288,13 @@ func (d *RktDriver) Validate(config map[string]interface{}) error {
 	fd := &fields.FieldData{
 		Raw: config,
 		Schema: map[string]*fields.FieldSchema{
+			// Image is only required if pod_manifest was not passed in
```
Comment (iaguis) [inline client/driver/rkt.go]: Image -> image

File: client/driver/rkt.go (hunk 21 above)
Comment (iaguis) [reply client/driver/rkt.go]: I agree, at least they should be consistent. That is, always be the same (either `img == ""` or `img != ""`. Maybe we can also define some meaningful variable name at the beginning, for example `podManifestSpecified := img == ""` so it's more clear.

File: client/driver/rkt.go [hunk 29]
```diff
@@ -563,18 +691,64 @@ func (d *RktDriver) Start(ctx *ExecContext, task *structs.Task) (*StartResponse,
...
+
+		if _, err := tmpfile.Write(podManifestStr); err != nil {
+			tmpfile.Close()
+			return nil, err
+		}
+		tmpfile.Close()
+
+		d.logger.Printf("[DEBUG] driver.rkt: wrote json to temp file %s", podManifestTmpfile)
+
+		// pod manifest is set here
```
Comment (iaguis) [inline client/driver/rkt.go]: To continue with the same comment style we can say:

```
// Set pod-manifest flag
```

File: client/driver/rkt.go (hunk 26 above)
Comment (iaguis) [reply client/driver/rkt.go:1036]: Please do this

File: helper/fields/data.go [hunk 30]
```diff
@@ -18,7 +18,20 @@ type FieldData struct {
...
 	var result *multierror.Error
 
-	// Scan for missing required fields
+	// Check that either image or pod_manifest was passed.
+	_, podManifestExists := d.Raw["pod_manifest"]
+	_, imageExists := d.Raw["image"]
+
+	if !podManifestExists && !imageExists {
+		result = multierror.Append(result, fmt.Errorf(
+			"both field pod_manifest and image are missing, but only one is required",
```
Comment (iaguis) [inline helper/fields/data.go:27]:
```go
result = multierror.Append(result, fmt.Errorf(`either field "pod_manifest" or "image" are required, but none was found`))
```

File: helper/fields/data.go [hunk 31]
```diff
@@ -18,7 +18,20 @@ type FieldData struct {
...
+		))
+	} else if podManifestExists && imageExists {
+		result = multierror.Append(result, fmt.Errorf(
+			"both field image and pod_manifest were passed, but only one is required"))
```
Comment (iaguis) [inline helper/fields/data.go]:
```go
result = multierror.Append(result, fmt.Errorf(`either field "pod_manifest" or "image" are required, but both were found`))
```

File: helper/fields/data.go (hunk 30 above)
Comment (iaguis) [reply helper/fields/data.go:27]: Won't this break other callers of `FieldData.Validate()`. Other drivers call it and might not have any of "pod_manifest" nor "image".

**Review by iaguis — CHANGES_REQUESTED** [review]
A first pass at this, I'll continue tomorrow.

Some general comments:

* This branch has conflicts, they need to be fixed
* I would put the vendor commit as the first one

File: helper/fields/data.go (hunk 30 above)
Comment (indradhanush) [reply helper/fields/data.go:27]: Good catch. But I'm not sure of a clean way to implement this check as the `Validate` method does not have driver information which would have helped to decide to do this check or not.

The other way I can think of is doing this check before calling the `Validate` function in `rkt.go` but that defeats the purpose of `Validate` and makes it hacky. Any thoughts on doing this in a better way?

File: client/driver/rkt.go (hunk 26 above)
Comment (indradhanush) [reply client/driver/rkt.go:1036]: Done, but doesn't reflect as outdated here because the lines after 1036 were changed.
